packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -0,0 +1,1136 @@
+  ///       });
+  ///     },
+  ///     tooltip: 'Reset',
+  ///     color: Theme.of(context).colorScheme.surface,
+  ///     icon: const Icon(Icons.replay),
+  ///   );
+  /// }
+  ///
+  /// @override
+  /// void initState() {
+  ///   super.initState();
+  ///   _controllerReset = AnimationController(
+  ///     vsync: this,
+  ///   );
+  /// }
+  ///
+  /// @override
+  /// void dispose() {
+  ///   _controllerReset.dispose();
+  ///   super.dispose();
+  /// }
+  ///
+  /// @override
+  /// Widget build(BuildContext context) {
+  ///   return Scaffold(
+  ///     backgroundColor: Theme.of(context).colorScheme.primary,
+  ///     appBar: AppBar(
+  ///       automaticallyImplyLeading: false,
+  ///       title: const Text('Controller demo'),
+  ///     ),
+  ///     body: Center(
+  ///       child: InteractiveViewer(
+  ///         boundaryMargin: EdgeInsets.all(double.infinity),
+  ///         disableRotation: true,
+  ///         transformationController: _transformationController,
+  ///         minScale: 0.1,
+  ///         maxScale: 1.0,
+  ///         onInteractionStart: _onInteractionStart,
+  ///         child: Container(
+  ///           color: Colors.pink,
+  ///         ),
+  ///       ),
+  ///     ),
+  ///     persistentFooterButtons: [resetButton],
+  ///   );
+  /// }
+  /// ```
+  /// {@end-tool}
+  ///
+  /// See also:
+  ///
+  ///  * [ValueNotifier].
+  ///  * [TextEditingController] for an example of another similar pattern.
+  final TransformationController transformationController;
+
+  @override _InteractiveViewerState createState() => _InteractiveViewerState();
+}
+
+class _InteractiveViewerState extends State<InteractiveViewer> with TickerProviderStateMixin {
+  final GlobalKey _childKey = GlobalKey();
+  Animation<Offset> _animation;
+  AnimationController _controller;
+  Offset _referenceFocalPoint; // Point where the current gesture began.
+  double _scaleStart; // Scale value at start of scaling gesture.
+  double _rotationStart = 0.0; // Rotation at start of rotation gesture.
+  double _currentRotation = 0.0; // Rotation of widget.transformationController.value.
+  BoxConstraints _constraints;
+  _GestureType _gestureType;
+
+  // This value was eyeballed as to give a feel similar to Google Photos.
+  static const double _kDrag = 0.0000135;
+
+  // The _boundaryRect is calculated by adding the boundaryMargin to the size of
+  // the child.
+  Rect _boundaryRectCached;
+  Rect get _boundaryRect {
+    if (_boundaryRectCached != null) {
+      return _boundaryRectCached;
+    }
+    assert(_childKey.currentContext != null);
+    assert(!widget.boundaryMargin.left.isNaN);
+    assert(!widget.boundaryMargin.right.isNaN);
+    assert(!widget.boundaryMargin.top.isNaN);
+    assert(!widget.boundaryMargin.bottom.isNaN);
+
+    final Size childSize = _getChildSize(
+      _childKey.currentContext.findRenderObject() as RenderBox,
+      _constraints,
+    );
+    _boundaryRectCached = Rect.fromLTRB(
```

Comment:
Not sure if it would be clearer, but you could write:
`_boundaryRectCached = boundaryMargin.inflateRect(Offset.zero & childSize);`

Reply:
I like it 👍